### DISCUSSION TITLE
CDMS2 support

### DISCUSSION THREAD

**Discussion by mauzey1** [PCMDI/cmor#688] [General]
I would like to start a discussion about when we should cut CDMS2 completely from CMOR 3.

[CDMS2 will not be supported beyond the end of 2023.](https://github.com/CDAT/cdms#cdms2)

The developers of CDMS2 are not planning on supporting Python 3.11 (https://github.com/conda-forge/cdtime-feedstock/pull/52#issuecomment-1302342070) so we are not testing with CDSM2 in our Python 3.11 builds.

A new issue has arose where [deprecated features in NumPy ](https://numpy.org/doc/stable/release/1.20.0-notes.html#deprecations)are no longer found in its [latest verison](https://numpy.org/doc/stable/release/1.24.0-notes.html#expired-deprecations).  Data type aliases `numpy.int` and `numpy.float` needed to be replaced with `int` and `float` in CMOR's Python wrapper to resolve #687.  However, the current build of CDMS2 from conda-forge still uses these deprecated aliases and will cause errors when used with NumPy 1.24.*

Here are some solutions that can resolve the above issue:
- **Remove CDMS2 support**:  This should be the easiest as it will just mean removing some lines from the Python wrappers, and changing a few parts of the testing suite.
- **Keep CMOR3's NumPy supported versions from 1.20 to 1.23**:  This will mean needing to restrict the NumPy version to 1.23 for the Conda environments used for our development and testing suite.  However, this restriction might be difficult to maintain in the conda-forge feedstock.
- **Get CDMS2 updated to use Numpy 1.24**:  This will resolve this issue upstream but is probably unlikely to happen given how CDMS2 support will end at the end of this year.

I would like to find a solution to this soon since it is currently a blocker for merge requests since they will fail at the CDMS2 tests.  I'm leaning more on removing CDMS2 support now rather than trying to restrict our NumPy version or waiting for the CDAT developers to update CDMS2.

**Comment by taylor13** [comment]
I think changing some parts of the testing suite is acceptable.  Are there other reasons that CDMS2 needs to be included in CMOR?  I think we need cdtime.  Is that part of CDMS2 or independent of it?

**Reply by mauzey1** [reply]
I could try setting the NumPy version to 1.23 for just the portion of the CI tests that use CDMS2.

CDMS2 is used by the Python API to access variables that were opened with CDMS2. 
```
    if numpy.ma.isMA(cell_bounds):
        cell_bounds = numpy.ascontiguousarray(cell_bounds.filled())
    elif has_oldma and numpy.oldnumeric.ma.isMA(cell_bounds):
        cell_bounds = numpy.ascontiguousarray(cell_bounds.filled())
    elif has_cdms2 and cdms2.isVariable(cell_bounds):
        cell_bounds = numpy.ascontiguousarray(cell_bounds.filled())
    elif has_cdms2 and cdms2.isVariable(cell_bounds):
        cell_bounds = numpy.ascontiguousarray(cell_bounds.filled())
    elif isinstance(cell_bounds, (list, tuple)):
        cell_bounds = numpy.ascontiguousarray(cell_bounds)
```
It's also used to get data such as units, time values, and time bounds.
```
    if numpy.ma.isMA(data):
        data = numpy.ascontiguousarray(data.filled())
    elif has_oldma and numpy.oldnumeric.ma.isMA(data):
        data = numpy.ascontiguousarray(data.filled())
    elif has_cdms2 and cdms2.isVariable(data):
        if time_vals is None:
            time_vals = data.getTime()
        data = numpy.ascontiguousarray(data.filled())
    elif isinstance(data, (list, tuple)):
        data = numpy.ascontiguousarray(data)
    elif not isinstance(data, numpy.ndarray):
        raise Exception("Error could not convert data to a numpy array")
```

CMOR's code already contains a copy of cdtime independent of the CDAT repo.
